lib/ddtrace/contrib/pg/connection.rb

```diff
@@ -0,0 +1,88 @@
+require 'ddtrace/ext/app_types'
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/sql'
+require 'ddtrace/contrib/pg/ext'
+
+module Datadog
+  module Contrib
+    module Pg
+      # PG::Connection patch module
+      module Connection
+        module_function
+
+        def included(base)
+          base.send(:prepend, InstanceMethods)
+        end
+
+        # PG::Connection patch instance methods
+        module InstanceMethods
+
+          # sync_exec(sql) -> PG::Result
+          # sync_exec(sql) {|pg_result| block}
+          def sync_exec(sql)
+            datadog_pin.tracer.trace(Ext::SPAN_QUERY) do |span|
+              span.resource = sql
+              span.service = datadog_pin.service
+              span.span_type = Datadog::Ext::SQL::TYPE
+              span.set_tag(Ext::TAG_DB_NAME, db)
+              span.set_tag(Datadog::Ext::NET::TARGET_HOST, host)
+              span.set_tag(Datadog::Ext::NET::TARGET_PORT, port)
+              super # this will pass all args, including the block
+            end
+          end
+
+          # sync_exec_params(sql, params[, result_format[, type_map]] ) -> PG::Result
+          # sync_exec_params(sql, params[, result_format[, type_map]] ) {|pg_result| block }
+          def sync_exec_params(sql)
+            datadog_pin.tracer.trace(Ext::SPAN_QUERY) do |span|
+              span.resource = sql
+              span.service = datadog_pin.service
+              span.span_type = Datadog::Ext::SQL::TYPE
+              span.set_tag(Ext::TAG_DB_NAME, db)
+              span.set_tag(Datadog::Ext::NET::TARGET_HOST, host)
+              span.set_tag(Datadog::Ext::NET::TARGET_PORT, port)
+              super # this will pass all args, including the block
+            end
+          end
+
+          # async_exec(sql) -> PG::Result OR
+          # async_exec(sql) {|pg_result| block}
+          def async_exec(sql)
```

Comment:
Is this for asynchronous queries to Postgres? If so, how does this work?

We have to be careful how it returns control back to Ruby, to make sure it doesn't close spans incorrectly or have problems resolving the correct trace context (which is currently bound by thread.)

Reply:
It uses the async API which responds to signals (vs. sync_exec doesn't respond to SIGTERM or SIGKILL). They suggest this over the synchronous mode so you don't lock up your processes entirely. It's not actually an asynchronous method where you have to call something else to get results (that exists, but I didn't instrument it)